implementations/rust/ockam/ockam/src/entity/profile.rs

```diff
@@ -31,40 +69,258 @@ impl Profile {
+
+        let change_event = ProfileChangeEvent::new(event_id, changes, vec![self_signature]);
+
+        let public_kid = v.sha256(public_key.as_ref())?;
+        let public_kid = ProfileIdentifier::from_hash(public_kid);
+
+        let profile = Profile::new(
+            public_kid,
+            ProfileChangeHistory::new(vec![change_event]),
+            Vec::new(),
+            vault.clone(),
+        );
+
+        Ok(profile)
+    }
+
+    pub fn create_key(
+        &mut self,
+        label: String,
+        key_type: ProfileKeyType,
+        key_purpose: ProfileKeyPurpose,
+        attributes: Option<ProfileEventAttributes>,
+    ) -> Result<(), ockam_core::Error> {
+        let event = self.create_key_event(label, key_type, key_purpose, attributes)?;
+        self.apply_no_verification(event)
+    }
+
+    pub fn rotate_key(
+        &mut self,
+        label: String,
+        key_type: ProfileKeyType,
+        key_purpose: ProfileKeyPurpose,
+        attributes: Option<ProfileEventAttributes>,
+    ) -> Result<(), ockam_core::Error> {
+        let event = self.rotate_key_event(label, key_type, key_purpose, attributes)?;
+        self.apply_no_verification(event)
+    }
+
+    pub fn get_private_key(
+        &self,
+        label: &str,
+        key_type: ProfileKeyType,
+        key_purpose: ProfileKeyPurpose,
+    ) -> Result<Secret, ockam_core::Error> {
+        let event = self
+            .find_last_key_event(label, key_type, key_purpose)
+            .unwrap(); // FIXME
+        let private_key = self
+            .get_private_key_from_event(label, key_type, key_purpose, event)
+            .unwrap(); // FIXME
+
+        Ok(private_key)
+    }
+}
+
+impl Profile {
+    fn check_consistency(_change_event: &ProfileChangeEvent) -> bool {
+        // TODO: check event for consistency: e.g. you cannot rotate the same key twice during one event
+        true
+    }
+
+    fn apply_no_verification(&mut self, change_event: ProfileChangeEvent) -> Result<(), Error> {
+        if !Self::check_consistency(&change_event) {
+            return Err(OckamError::InvalidInternalState.into());
         }
 
-        for _change in change_event.changes() {
-            // TODO: apply change
-            unimplemented!()
+        self.change_history.push_event(change_event);
+
+        Ok(())
+    }
+
+    pub fn apply(&mut self, change_event: ProfileChangeEvent) -> Result<(), Error> {
+        if !self.verify(&change_event)? {
+            return Err(OckamError::VerifyFailed.into());
         }
+
+        self.apply_no_verification(change_event)
     }
 
-    pub fn verify(&self, _change_event: &ProfileChangeEvent) -> bool {
-        // loop over all proofs and verify them
-        unimplemented!()
+    // WARNING: Checks only one event, assumes all previous events are verified
+    pub fn verify(&self, change_event: &ProfileChangeEvent) -> Result<bool, ockam_core::Error> {
```

Comment:
I think this function can be changed to return an error vs false. It should simplify your logic.